Make rawrtc compile on ARM by making SCTP redirect transport optional

JesApp opened the pull request:
This pull request enables rawrtc to be cross-compiled for ARM processors by conditionally removing the intel-specific code in file `src/librawrtc/crc32c.c`.

# The problem
Rawrtc can not be compiled for ARM processors because the file `crc32c.c` uses intel-specific code to speed up the calculation. While a software fallback is provided, it is only selected at runtime which does not help with compiling. (See file `crc32c.c` for an explanation.)

# Suggested solution
Remove the intel-specific code and replace it with the already existing software solution at compile time.

## Changes

1. A [CMake toolchain](https://cmake.org/Wiki/CMake_Cross_Compiling) file `gcc-arm.cmake` is provided that sets the C and C++ compilers to cross-compilers from the gcc suite as well as the `CMAKE_SYSTEM_PROCESSOR` variable to `arm`.
1. `src/librawrtc/CMakeLists.txt` is edited to set a preprocessor variable `TARGET_NOT_X86` if `CMAKE_SYSTEM_PROCESSOR` is `arm`.
1. ` src/librawrtc/crc32c.c` is edited to compile only the software version if `TARGET_NOT_X86` is set.

# Open questions
1. This should be documented in some way. New section in Readme? New file? Comments in the toolchain file?
1. The toolchain file is probably fairly specific to my situation. Should we leave that in as an example? Maybe make it more generic?

# Notes
* Work in progress because while this compiles, I haven't actually tested the result yet.
* I'm not an expert when it comes to CMake. If there's a better way of doing this, let me know.
* I can provide a docker container where this compiles, if needed. It's fairly specific to the embedded system I'm using but if anyone would like to try this out it should be good enough for that.

Side note: I'm on holiday until August 27th, so no rush on this.

lgrahl commented:
Hey @JesApp, that's a neat PR so far! One suggestion: Instead of adding software CRC32, I'd propose to just disable (e.g. leave it out from compilation) the SCTP redirect transport on ARM. This tool will be used rather rarely and mostly just for testing purposes. So, my proposal is this:

1. Add a switch to disable SCTP redirect transport (this is the only place where CRC32 is being needed).
2. Force enable that switch in case architecture is ARM-based.

What do you think?

lgrahl commented:
Correction: I think the SCTP redirect transport should be disabled by default. So, the switch should enable the redirect transport (and the CRC32 code).

Note: usrsctp also has (software) CRC32 but in the future there might be a flag to disable data channels, so a hard dependency on that isn't useful.

JesApp commented:
Sure, that works for me. I'll make those changes sometime this week.

JesApp commented:
I have added a CMake variable and changed the compile target depending on it. However, the build now fails [with these messages](https://github.com/rawrtc/rawrtc/files/1256803/sctp_redirect_link_fail.txt).

I will look into this further tommorow.

lgrahl commented on gcc-arm.cmake:
```diff
@@ -0,0 +1,13 @@
+#target system is a linux
+set (CMAKE_SYSTEM_NAME Linux)
+#target processor is an arm with a version number <8
+#this is relevant since the only problematic part of rawrtc is the intel-specific assembler code in crc32c.c and arm supports its own assembler instructions since version 8
```
This line should only be needed in case the redirect tool is being compiled, correct? In that case, it should probably be if-def'ed, too.

JesApp replied:
If I get this working the way I plan to, the only lines that will be needed are the ones that set CC and CXX. That can be done just as well via environment variables, so I'll (hopefully) delete `gcc-arm.cmake`.

JesApp commented:
Allright, I added a cached cmake variable called `SCTP_REDIRECT_TRANSPORT` that's set to `OFF`by default. If it's `OFF`, `sctp_redirect_transport.c` and `crc32c.c` do not get added to the sources and the parts you mentioned above get removed from the source code by the preprocessor.

As I wrote above, I have also removed the toolchain file.

This works for me, both on intel and cross-compiling for ARM.

JesApp commented:
I've just added some information about the new flag to the readme.

As far as I'm concerned, this is done. If you'd like any changes, let me know.

lgrahl commented on Readme.md:
```diff
@@ -229,6 +229,16 @@ This tool has been developed to be able to test data channel implementations
 
+Building:
+
+This tool is not built by default. In order to build it, set the cached CMake flag `SCTP_REDIRECT_TRANSPORT` to `ON` when building, like so: 
```
I'd replace *cached CMake flag* with *environment variable* as the internals don't really matter to the user.

lgrahl commented on Readme.md:
```diff
@@ -229,6 +229,16 @@ This tool has been developed to be able to test data channel implementations
+    cmake -DCMAKE_INSTALL_PREFIX=${PWD}/prefix -DSCTP_REDIRECT_TRANSPORT=ON ..
+    make install
+    
+Note that this tool contains intel-specific code and will not build for non-intel system like ARM.
```
While stating *Intel-specific* is correct, it's a bit misleading. I'd propose saying *[...] will not build without SSE 4.2 support.* instead.

lgrahl commented on src/librawrtc/CMakeLists.txt:
```diff
@@ -15,11 +14,19 @@ set(rawrtc_SOURCES
 
+#if we are building the SCTP redirect transport tool
+if(SCTP_REDIRECT_TRANSPORT)
+  #add the relevant source files
```
Missing space and capitalise.

lgrahl commented on src/librawrtc/CMakeLists.txt:
```diff
@@ -30,6 +37,12 @@ target_link_libraries(rawrtc-static
         DESTINATION lib)
 
+if(SCTP_REDIRECT_TRANSPORT)
+  #and set the preprocessor variable
```
Missing space and capitalise.

lgrahl commented on src/tools/CMakeLists.txt:
```diff
@@ -29,13 +29,18 @@ install(TARGETS dtls-transport-loopback
-        rawrtc-helper)
-install(TARGETS sctp-redirect-transport
-        DESTINATION bin)
+
```
The comment should stay just above the block (remove the empty line).

lgrahl commented on CMakeLists.txt:
```diff
@@ -26,6 +26,9 @@ if (${CMAKE_BUILD_TYPE} MATCHES "DEBUG")
 endif ()
 
+# Flag for enabling building of SCTP redirect transport
+set (SCTP_REDIRECT_TRANSPORT OFF CACHE BOOL "Build the SCTP redirect transport tool.")
```
Please don't use a space here after `set`.

lgrahl commented on src/librawrtc/CMakeLists.txt:
```diff
@@ -15,11 +14,19 @@ set(rawrtc_SOURCES
         sctp_transport.c
         utils.c)
 
+#if we are building the SCTP redirect transport tool
```
Please add a space after `#` and capitalise the first word.

lgrahl commented on src/librawrtc/CMakeLists.txt:
```diff
@@ -15,11 +14,19 @@ set(rawrtc_SOURCES
+        crc32c.c
+        sctp_redirect_transport.c
+    )
+endif()
```
`endif ()` rather than `endif()` (coding convention).

lgrahl commented on src/librawrtc/CMakeLists.txt:
```diff
@@ -15,11 +14,19 @@ set(rawrtc_SOURCES
         utils.c)
 
+#if we are building the SCTP redirect transport tool
+if(SCTP_REDIRECT_TRANSPORT)
```
`if (...)` rather than `if(...)` (coding convention)..

lgrahl commented on src/tools/CMakeLists.txt:
```diff
@@ -29,13 +29,18 @@ install(TARGETS dtls-transport-loopback
-        DESTINATION bin)
+
+if (SCTP_REDIRECT_TRANSPORT)
+  message(STATUS "Configuring for build of SCTP redirect transport.")
```
*of* -> *with*

lgrahl commented on src/tools/helper/CMakeLists.txt:
```diff
@@ -9,3 +9,6 @@ set(rawrtc_HELPER
         rawrtc)
+if (SCTP_REDIRECT_TRANSPORT)
+  target_compile_definitions(rawrtc-helper PRIVATE SCTP_REDIRECT_TRANSPORT)
+endif()
```
`endif ()` rather than `endif()` (coding convention).

lgrahl commented on src/tools/CMakeLists.txt:
```diff
@@ -29,13 +29,18 @@ install(TARGETS dtls-transport-loopback
+  install(TARGETS sctp-redirect-transport
+          DESTINATION bin)
+  target_compile_definitions(sctp-redirect-transport PRIVATE SCTP_REDIRECT_TRANSPORT)
+endif()
```
`endif ()` rather than `endif()` (coding convention).

JesApp commented:
Thanks for the feedback. :) I made the changes you requested.
